espresso/authdb/authdb.go

```
@@ -186,77 +218,194 @@ func (op *AuthAncientReaderOp) AncientSize(kind string) (uint64, error) {
+		for _, tw := range op.tagBatch {
+			tagTable, ok := getTagTable(tw.kind)
+			if !ok {
+				continue // Skip unsupported tables
+			}
+			if err := tagOp.AppendRaw(tagTable, tw.number, tw.tag); err != nil {
+				return fmt.Errorf("failed to write tag for %s#%d: %w", tw.kind, tw.number, err)
+			}
+		}
+		return nil
+	})
+	return err
 }
 
 func (d *AuthDB) ModifyAncients(fn func(ethdb.AncientWriteOp) error) (int64, error) {
-	authFn := func(op ethdb.AncientWriteOp) error {
-		authOp := AuthAncientWriteOp{
-			inner: op,
-			mac:   d.mac,
+	if d.mac == nil {
+		return d.db.ModifyAncients(fn)
+	}
+
+	// Create our wrapper that collects tag writes
+	var authOp *AuthAncientWriteOp
+
+	// Execute main writes
+	writeSize, err := d.db.ModifyAncients(func(op ethdb.AncientWriteOp) error {
+		authOp = &AuthAncientWriteOp{
+			inner:    op,
+			authDB:   d,
+			tagBatch: make([]tagWrite, 0),
 		}
 		return fn(authOp)
+	})
+
+	// Main write failed - don't write tags
+	if err != nil {
+		return writeSize, err
+	}
+
+	// Main writes succeeded - now write tags
+	if authOp != nil {
+		if tagErr := authOp.writeTags(); tagErr != nil {
+			// Future reads will fail verification
+			log.Crit("Failed to write authentication tags after successful main write",
+				"error", tagErr, "writes", len(authOp.tagBatch))
+			return writeSize, fmt.Errorf("tag write failed: %w", tagErr)
+		}
 	}
-	return d.db.ModifyAncients(authFn)
+
+	return writeSize, nil
 }
 
 func (d *AuthDB) TruncateHead(n uint64) (uint64, error) {
-	return d.db.TruncateHead(n)
+	old, err := d.db.TruncateHead(n)
+	if err != nil {
+		return old, err
+	}
+
+	// Truncate tag freezer to match
+	if d.mac != nil && d.tagFreezer != nil {
+		// Truncate all tag tables to the same head, but only if they have data
+		for _, tagTable := range freezerTabletoTagTable {
+			// Check if table has any items before truncating
+			if items, _ := d.tagFreezer.Ancients(); items > 0 {
```

What if items is less than n?

---

it would fail right? just as you would in a normal `old, err := d.Database.TruncateHead(n)`. Do you have some concern about this? 

I will try to NOT ignore err and log.Crit on failure since the ancient might enter a limbo state (where the data is truncated but not the tag, they will be inconsistent but manually fixable, not the end of the world, but need to be apparent)